降順昇順切り替え機能【受講生一覧画面】※講師側（DAIKI）

i-k-i-a-d opened the pull request:
やったこと
・StudentController内に生徒の一覧をカラム名によって昇順と降順で並び替えるメソッドを作成。
・Seederファイルを修正し、attendancesテーブルとlesson_attendancesテーブルにテストデータを追加。 
・新しく作成した並び替え用のメソッドまでのルーティングを実装。
・並び替えメソッド用のバリデーションルールを作成。
・バリデーションルールを細かく記述するために二つのルールクラスを作成。

確認方法
Postmanで
http://localhost:8080/api/v1/instructor/course/1/student/{column}/{order}/indexをSend。
※{column}にはnick_name,email,title,created_at,last_login_atから選択して入力
※{order}には昇順の場合asc、降順の場合descを入力

すると正常に生徒の情報が並び変わって返ってくる。

i-k-i-a-d committed:
StudentController内に生徒の一覧をカラム名によって昇順と降順で並び替えるメソッドを作成。
Seederファイルを修正し、attendancesテーブルとlesson_attendancesテーブルにテストデータを追加。
新しく作成した並び替え用のメソッドまでのルーティングを実装。
並び替えメソッド用のバリデーションルールを作成。
バリデーションルールを細かく記述するために二つのルールクラスを作成。

kou-tech commented on app/Http/Controllers/Api/Instructor/StudentController.php:
既に一覧取得APIがあるので、そこにソート機能を実装するようにお願いします。

i-k-i-a-d replied:
要件定義を見誤っておりました。

kou-tech commented on app/Http/Requests/Instructor/SortStudentsRequest.php:
```diff
@@ -0,0 +1,45 @@
+            'course_id' => ['required', 'integer', 'exists:courses,id'],
+            'per_page' => ['integer', 'min:1'],
+            'page' => ['integer', 'min:1'],
+            'column' => ['required', 'string', new SortStudentsColumnRule],
+            'order' => ['required', 'string', new SortStudentsOrderRule],
```
APIにとって、ソートする対象を`column`とするのは、少し違和感を感じるので、sortByなどいかがでしょう？

kou-tech replied:
orderは汎用的な値でStudentに限らない値なので、次のように書いてしまっても良いかもしれません。

`'in:asc,desc'`

判断軸は難しいですが、このケースの場合のカスタムルールはやりすぎな実装になるかもという話でした！

kou-tech replied:
`column`と`order`の必須条件はなくし、コントローラーで値がない場合のデフォルト値を設定するようにお願いします。

i-k-i-a-d replied:
修正しました。

kou-tech commented on app/Model/Attendance.php:
```diff
@@ -54,4 +54,15 @@ public function lessonAttendances()
     }
+
+    //sortStudentsメソッドで受け取るパラメータのカラム名を定義
+    const NICK_NAME_COLUMN = 'nick_name';
+    const EMAIL_COLUMN = 'nick_name';
+    const TITLE_COLUMN = 'nick_name';
+    const CREATED_AT_COLUMN = 'nick_name';
+    const LAST_LOGIN_AT_COLUMN = 'nick_name';
```
これは私がそうしているのですが、先頭で揃えて定義しておくと可読性が良いと思います！
しかし、現場のルールによります！

````php
    const COLUMN_NICK_NAME = 'nick_name';
    const COLUMN_EMAIL = 'nick_name';
    const COLUMN_TITLE = 'nick_name';
    const COLUMN_CREATED_AT = 'nick_name';
    const COLUMN_LAST_LOGIN_AT = 'nick_name';
````

kou-tech replied:
あと全部同じ値になっていますので、修正をお願いします。

i-k-i-a-d replied:
確かに見やすいですね。今後はこのように記述します。
また、これは完全にうっかりでした。

kou-tech commented on app/Http/Controllers/Api/Instructor/StudentController.php:
```diff
@@ -72,4 +74,40 @@ public function store(StudentStoreRequest $request)
+            ], 403);
+        }
+
+        if ($request->column !== 'title') {
+            $attendances = Attendance::with(['student', 'course'])
+                                        ->where('course_id', $request->course_id)
+                                        ->join('students', 'attendances.student_id', '=', 'students.id')
+                                        ->orderBy('students.' . $request->column, $request->order)
+                                        ->paginate($perPage, ['*'], 'page', $page);
+        } else {
+            $attendances = Attendance::with(['student', 'course'])
+                    ->where('course_id', $request->course_id)
+                    ->join('courses', 'attendances.course_id', '=', 'courses.id')
+                    ->orderBy('courses.' . $request->column, $request->order)
+                    ->paginate($perPage, ['*'], 'page', $page);
+        }
```
MTGで話した通りこれだと似たようなコードが重複してしまうので、whenを活用した方が良いと思います。

コード上でTitleとTitle以外のようなコードを書くと今後受講テーブルの他のカラムでソートする場合などの拡張性に欠けるので、in_arrayなどを使うと良いです！

次のようなイメージです。

````php
        $attendances = Attendance::with(['student', 'course'])
                                    ->where('course_id', $request->course_id)
                                    ->when(
                                        in_array($sortBy, [Attendance::COLUMN_NICK_NAME, Attendance::COLUMN_EMAIL, Attendance::COLUMN_LAST_LOGIN_AT], true),
                                         function ($query) use ($sortBy, $order) {
                                        // 生徒テーブルのカラムでソートするケース
                                    })
                                    ->when(
                                        in_array($sortBy, [Attendance::COLUMN_TITLE], true),
                                         function ($query) use ($sortBy, $order) {
                                        // 講座テーブルのカラムでソートするケース
                                    })
                                    ->orderBy($sortBy, $order)
                                    ->paginate($perPage, ['*'], 'page', $page);
````

kou-tech replied:
より可読性を上げる方法に比較条件を関数式に置き換える方法もあるので、覚えておいてください。

````php
     ->when($this->isCourseColumn($sortBy),
     ...

    /**
     * 講座テーブルのカラム名かどうかを判定
     */
    private function isCourseColumn(string $sortBy): bool
    {
        return in_array($sortBy, [Attendance::COLUMN_TITLE], true);
    }
````

i-k-i-a-d replied:
whenメソッドの第一引数の条件式を関数式にして実装済みです。

kou-tech commented on app/Http/Requests/Instructor/SortStudentsRequest.php:
```diff
@@ -0,0 +1,45 @@
+    {
+        $this->merge([
+            'course_id' => $this->route('course_id'),
+            'column' => $this->route('column'),
+            'order' => $this->route('order'),
```
URLではなく、クエリで送信するようにお願いします！

i-k-i-a-d replied:
formからPOSTではなくGETで送信することをクエリと呼ぶんですね。知りませんでした。

i-k-i-a-d committed:
indexメソッドに昇順と降順の並べ替え機能を移植。
indexメソッド用のバリデーションルールを修正。
sortStudentsメソッドを削除し、そのバリデーションルール用のリクエストクラスも削除。
sortStudentsメソッド用のリクエストクラス用のルールファイル2つをindexメソッド用にリネーム。
sortStudentsメソッドの用のルーティングを削除。

kou-tech commented on app/Http/Controllers/Api/Instructor/StudentController.php:
```diff
@@ -6,13 +6,16 @@
 use App\Http\Resources\Instructor\StudentShowResource;
 use App\Http\Requests\Instructor\StudentStoreRequest;
 use App\Http\Resources\Instructor\StudentStoreResource;
+use App\Http\Requests\Instructor\SortStudentsRequest;
```
こちらは不要なので削除お願いします。

i-k-i-a-d replied:
既にファイル自体がないので削除しました。

kou-tech commented on app/Http/Requests/Instructor/StudentIndexRequest.php:
```diff
@@ -32,9 +34,11 @@ public function authorize()
             'per_page' => ['integer', 'min:1'],
             'page' => ['integer', 'min:1'],
+            'sortBy' => ['string', new IndexSortByRule],
+            'in:asc,desc' => ['string', new IndexOrderRule],
```
こちら誤った認識されていますね。
次のようにして欲しいという内容で、定数で管理はやりすぎ(=冗長)という話でした。

````php
            'order' => ['string', 'in:asc,desc'],
````

i-k-i-a-d replied:
これは恥ずかしいですね。
あとこういうバリデーションの記述方法があるのを知りませんでした。
このように記述すると、orderにはascかdescしか入らなくなるんでしょうか？

kou-tech replied:
そうです！
ただし、レビューする都合上、私も動作確認していますがご自身でも動作確認は必ずしてください！

![スクリーンショット 2023-11-12 13 04 58](https://github.com/yukihiroLaravel/juko_laravel/assets/64859506/f6180868-eafd-4e3a-bc93-6047c2edddb4)

i-k-i-a-d replied:
このエラー確認済みだったんですが、メッセージの内容がascかdescにしてください的な内容ではなかったので、これで正しい挙動なのかどうか不明だったので質問いたしました。

kou-tech commented on app/Model/Attendance.php:
```diff
@@ -54,4 +54,15 @@ public function lessonAttendances()
+    const COLUMN_EMAIL = 'email';
+    const COLUMN_TITLE = 'title';
+    const COLUMN_CREATED_AT = 'created_at';
+    const COLUMN_LAST_LOGIN_AT = 'last_login_at';
```
内容が誤っているコメントの修正と`CLOUMN`は変更されたと思うので、`SORT_BY`に変更をお願いします。

````php
const SORT_BY_NICK_NAME = 'nick_name';
const SORT_BY_EMAIL = 'email';
const SORT_BY_TITLE = 'title';
const SORT_BY_CREATED_AT = 'created_at';
const SORT_BY_LAST_LOGIN_AT = 'last_login_at';
````

i-k-i-a-d replied:
実質カラム名なので、カラムの方がわかりやすいかと、そのままにしてました。

kou-tech replied:
> 実質カラム名

ソート対象のカラム名が適した文脈である認識ですよね！
他の開発者を想定して欲しいのですが、どうしてわざわざカラムの定数を定義しているかを`CLOUMN_`だけだと把握が難しいですよね。コメントを書けばいいと思うかもしれませんが、基本的にコメントはメンテされずに負債化となりがちなので、でしたらなるべく定数名でどういうための定数なのかを表現したいという話でした！

kou-tech commented on app/Http/Controllers/Api/Instructor/StudentController.php:
```diff
@@ -26,10 +29,28 @@ public function index(StudentIndexRequest $request)
         $perPage = $request->input('per_page', 20);
         $page = $request->input('page', 1);
+        $sortBy = $request->input('sortBy', 'nick_name');
+        $ascOrDesc = $request->input('in:asc,desc', 'asc');
```
ここはorderのままで良いと思います！

i-k-i-a-d replied:
変数名のさじ加減が難しいですね。

kou-tech replied:
変数名は何を示すデータなのかを意識すると良いかもしれません！

「$ascOrDesc」は変数名からは何を表しているのかが明確ではないですよね。一方、「$order」は、変数名から何を表しているのかが明確であり、コードの可読性を向上させます。

ただし、「$order」より適切な変数名の考慮はあると考えています。

i-k-i-a-d replied:
中身がascかdescなので、これが一番わかりやすいと思って採用したんですが、逆にわかりづらかったみたいで申し訳ありません。

kou-tech replied:
全然謝る必要はないです！
わかりづらい訳ではないのですが、このケースにおいては不適切という話でした！

kou-tech commented on app/Http/Controllers/Api/Instructor/StudentController.php:
```diff
@@ -26,10 +29,28 @@ public function index(StudentIndexRequest $request)
+        $loginId = Auth::guard('instructor')->user()->id;
+        $instructorId = Course::findOrFail($request->course_id)->instructor_id;
+
+        if ($loginId !== (int)$instructorId) {
```
Courseモデルにキャスト変換を記述して、型変換した方が良いです！

i-k-i-a-d replied:
ここも指摘される気はしてました。

kou-tech commented on app/Http/Controllers/Api/Instructor/StudentController.php:
```diff
@@ -26,10 +29,28 @@ public function index(StudentIndexRequest $request)
+            ->when($this->isCourseColumn($sortBy), function (Builder $query) {
+                $query->join('courses', 'attendances.course_id', '=', 'courses.id');
+            })
+            ->orderBy($sortBy, $ascOrDesc)
```
OrderByするテーブルが違うので、この記述もwhenの中に含めるべきでないですか？
私の認識が間違っていたらすみません！

i-k-i-a-d replied:
これについては、僕もよくわかっていないのですが、このままでも問題なく動いてます。
おそらくwhenの特徴として第一引数の条件式がfalseのときは何処理が起こらずスルーされるのと、最初のwhenの状況と2回目のwhenの状況が同時に起きないので片方で、並べ替えたときは、もう片方が並べ変わらないので、そのまま最後にorderByすれば、片方のテーブルとだけ連結した状態でまとめて並び替えることができるのだと思われます。
飽くまで個人的見解です。

よって、今回は修正しておりませんが、お気づきの点があればご指摘ください。

あと、そもそもですが、titleについては、講座名で並び替えるようにという要件定義ですが、このAPIはURLで講座idを指定しているので、それに紐づく講座名は一択となり、全く同じ講座名で昇順や降順で並び替えるということ自体が意味をなさない気がします。

kou-tech replied:
> 全く同じ講座名で昇順や降順で並び替えるということ自体が意味をなさない気がします。

そうですね！また、十分に動作確認も難しいと思われるため、講座名のソートは含めないように提案してみます！

i-k-i-a-d replied:
作業会でユキヒロさんにも直接確認し、今回はtitleでのソートは見送ることになったので、titleでソートをかけれるように記述していたロジックについては全て削除しました。

i-k-i-a-d committed:
StudentControllerから不要なインポートを削除
StudentController内のindexメソッドの変数名、inputメソッドの引数、orderByメソッドの引数、型指定のintの削除
Requestクラスのバリデーションルールを一部修正。
Attendanceモデル内にでinstructor_idの型変換を実装
上記訂正、AttendanceモデルではなくCourseモデル
Attendanceモデル内の定数を一部削除、一部定数名を修正。
やりすぎだったので、Ruleクラスを削除。
Ruleクラス内の定数名を修正。

kou-tech commented on app/Model/Attendance.php:
```diff
@@ -54,4 +54,12 @@ public function lessonAttendances()
         return $this->hasMany(LessonAttendance::class);
     }
+
+    //indexメソッドで受講生の一覧を並び替える際にクエリで受け取る値を定義
```
こちらのコメントは、今後indexメソッドの改修のたびに更新を考慮しなくてはならないので、不要です！
残すとしても、`// ソート項目`程度に収めておくと良いです！

i-k-i-a-d replied:
修正しました。コメントって難しいですね。

kou-tech commented on app/Rules/IndexSortByRule.php:
```diff
@@ -0,0 +1,56 @@
+     */
+    public function message()
+    {
+        return 'Invalid Colunm Name.';
```
Sort By にしておきましょう！

i-k-i-a-d replied:
修正しました。

i-k-i-a-d committed:
Attendanceモデル内で定数定義した際のコメントを修正。
IndexSortByRuleクラスで定義したバリデーションエラー時のコメントを修正。
titleに関するソートは見送りとなったので、titleでソートするための記述を全て削除。

kou-tech commented on app/Http/Controllers/Api/Instructor/StudentController.php:
```diff
@@ -26,10 +28,23 @@ public function index(StudentIndexRequest $request)
+            ], 403);
+        }
+
+        $attendances = Attendance::with(['student'])
```
リソースファイル内で次の記述があるので、withにcourseは必要でした！
以前の実装がEager Loadingを考慮できていなかったですね。。。

````php
$attendance->course->title
````

i-k-i-a-d replied:
失礼しました。Resourceクラスの戻り値をきちんと確認していれば防げることでした。申し訳ありません。
今後の改修を考慮してあえて、Eager Loadingを行っていなかったと理解しました。
複数の講座を対象として生徒情報を取得する改修、、、、やってみたかったです。

i-k-i-a-d committed:
やっぱり必要だとわかったので、Attendanceモデルのwithメソッドで引数にcourseモデルを文字列で渡すように修正。
